hic.wdl

```diff
@@ -58,207 +49,156 @@ workflow hic {
-        call make_restriction_site_locations { input:
-            reference_fasta = select_first([reference_fasta]),
-            assembly_name = select_first([assembly_name, "unknown_assembly"]),
-            restriction_enzyme = restriction_enzymes[0],
+        if (!is_nonspecific && !defined(restriction_sites)) {
+            call exit_early { input:
+                message = "Must provide restriction sites file if enzyme is not `none`"
+            }
         }
     }
 
-    Boolean has_restriction_sites = defined(restriction_sites) || defined(make_restriction_site_locations.restriction_site_locations)
-
     # scatter over libraries
-    if (has_restriction_sites && defined(chrsz) && !restriction_site_locations_only) {
-        File chrsz_ = select_first([chrsz])
-        File restriction_sites_ = if defined(restriction_sites) then select_first([restriction_sites]) else select_first([make_restriction_site_locations.restriction_site_locations])
-        Int num_bioreps = if defined(bams) then length(select_first([bams])) else length(fastq)
-
-        scatter(i in range(num_bioreps)) {
-            # Align fastqs if input bams were not provided
-            if (!defined(bams) && defined(reference_index)) {
-                scatter(j in range(length(fastq[i]))) {
-                    call align { input:
-                        fastqs = fastq[i][j],
-                        chrsz = chrsz_,
-                        idx_tar = select_first([reference_index]),
-                        ligation_site = ligation_site,
-                        num_cpus = align_num_cpus,
-                    }
-                    if (length(read_groups) > 0) {
-                        call add_read_group_to_bam { input:
-                            bam = align.result,
-                            read_group = read_groups[i][j],
-                        }
+    Int num_bioreps = if defined(bams) then length(select_first([bams])) else length(fastq)
+
+    scatter(i in range(num_bioreps)) {
+        # Align fastqs if input bams were not provided
+        if (!defined(bams) && defined(reference_index)) {
+            scatter(j in range(length(fastq[i]))) {
+                call align { input:
+                    fastqs = fastq[i][j],
+                    idx_tar = select_first([reference_index]),
+                    ligation_site = select_first([ligation_site]),
+                    num_cpus = align_num_cpus,
+                }
+                if (length(read_groups) > 0) {
+                    call add_read_group_to_bam { input:
+                        bam = align.bam,
+                        read_group = read_groups[i][j],
                     }
                 }
-                Array[File] aligned_bams = select_all(if length(read_groups) > 0 then add_read_group_to_bam.bam_with_read_group else align.result)
             }
+            Array[File] aligned_bams = select_all(if length(read_groups) > 0 then add_read_group_to_bam.bam_with_read_group else align.bam)
```

Comment:
Just a suggestion. WDL 1.0 allows `scatter` with an empty array. So we can have a separate `scatter` loop for `fastqs` and simplify the declaration of (or even remove) `num_bioreps`.
```
scatter(i in fastqs) {
  if( defined(reference_index ) {
     call align {
     }
  }
}
Array[File] aligned_bams = flatten([align.bam, bams])

scatter(bam in aligned_bams ) {
}
```

Reply:
Nice. I just tried that method and it's definitely promising, although I think it will require some refactoring that is beyond the scope of this ticket.